[bnb] Let's make serialization of 4bit models possible

poedator opened the pull request:
## What does this PR do?
Purpose: enable saving and loading transformers models in 4bit formats.

tested with Bloom-560 and Llama-2-7b. Save - Load - match tensors and quant_states - match inference results.

## connection with bitsandbytes
Requires this PR in bitsandbytes: https://github.com/TimDettmers/bitsandbytes/pull/753 to be able to save/load models

## testing:
the functionality was tested doing this series of commands:
```
model = m4 = transformers.AutoModelForCausalLM.from_pretrained(..., [with 4bit quant config])
model.save_pretrained(SAVE_PATH, safe_serialization= [False / True] )
model2 = transformers.AutoModelForCausalLM.from_pretrained( SAVE_PATH)
# then matching all params and quant_state items between the models, matching inference results
```
Specific tests will be added to this PR once the [bitsandbytes PR](https://github.com/TimDettmers/bitsandbytes/pull/753) merges.

## Open questions to the maintainers:
1. I added and tested code necessary for straightforward save/load operations with LLMs. Yet there may be other kinds of function calls that may need to be updated to handle 4bit save/load - pls suggest if/where to expand this PR.
specifically: this PR covers `load_state_dict_into_meta_model()` but not `load_state_dict_into_model()` for I did not find an  example that uses it. 

2. Some of my edits may not fit with style / refactoring plans of the maintainers - pls give guidance if needed.

amyeroberts commented:
cc @younesbelkada @SunMarc

poedator commented:
## on saving format for quantized weight.
Right now there is no clear separation of code to load a pretrained model in fp16/fp32 with subsequent quantisation and loading already quantized model. I noticed that for 8-bit it comes down to detecting `SCB` in state_dict.keys(). In this PR the 4bit quantized param storage requires not one but about 10 components per weight to be loaded from the state_dict - it may become unpractical.
Also saving all quantization params as _buffers into state_dict() may be cluttering the state_dict, and be impractical for some data types.
This matter is important right now, because of multitude of quantization methods (and more to come): int8, nd4, fp4 in bnb, SpQR, GPTQ, etc. Some guidelines here will simplify transformers code and facilitate adding new quantizations later. 
One possibility is this:
Packing all of the quantisation components into single serializeable object per parameter (protobuf?) and then saving it alongside the main parameter data. Then the loading code may look like:
```
for k in state_dict:
    if k + '_quant_state' in loaded_state_dict:
        new_param_weight = load_quantized_weight(loaded_state_dict[k], loaded_state_dict[k+'_quant_state'], )
    else:
        new_param_weight = loaded_state_dict[k]
...
def unpack_quantized(data, packed_quant_state):
    quant_type, quant_state = unpack_q_state(packed_quant_state)
    if quant_type == 'bnb_nf4':
        new_param_weight = bnb.nn.Param4bit.from_prequantized(data, quant_state) # call specific unpacker
    ...
    return new_param_weight  # fully unpacked and loaded unto cuda

def unpack_q_state(packed_quant_state: protobuf) -> Tuple(str, SomeQuantStateStru):
    """from serialized format to dict(or some data structure) of quantization state"""
    ...
    return quant_type, quant_state
```
Alternatively, quant_type per parameter may be stored in `config.json`, but it becomes too fat then.
I suggest to allow various quantization types per parameter.
What do you think? I'll be glad to refactor this PR to make it future-proof. 
@younesbelkada @SunMarc @TimDettmers

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,13 +76,14 @@ class `Int8Params` from `bitsandbytes`.
                 new_value = value.to("cpu")
-                if value.dtype == torch.int8:
+                if value.dtype in (torch.int8, torch.uint8):
                     is_8bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse(
                         "0.37.2"
                     )
```
I think we should check for the correct bnb version when it comes to 4bit as well, if the PR https://github.com/TimDettmers/bitsandbytes/pull/753 gets merged, it should probably be greater than the current `0.41.1` version

poedator replied:
added checks with `is_4bit_serializable `

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,13 +76,14 @@ class `Int8Params` from `bitsandbytes`.
                         "0.37.2"
                     )
                     if not is_8bit_serializable:
+                        bits_ = 4 if is_4bit else 8
                         raise ValueError(
-                            "Detected int8 weights but the version of bitsandbytes is not compatible with int8 serialization. "
+                            f"Detected {bits_} weights but the version of bitsandbytes is not compatible with {bits_} bit serialization. "
                             "Make sure to download the latest `bitsandbytes` version. `pip install --upgrade bitsandbytes`."
                         )
```
Then here we would break it down in two cases -> 8bit serializable through `is_8bit_serializable` and 4-bit serializable through a new bool `is_4bit_serializable`

poedator replied:
Done. Separated 2 cases more clearly.

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -95,7 +98,14 @@ class `Int8Params` from `bitsandbytes`.
                 new_value = bnb.nn.Int8Params(new_value, requires_grad=False, **kwargs).to(device)
             elif is_4bit:
-                new_value = bnb.nn.Params4bit(new_value, requires_grad=False, **kwargs).to(device)
+                if new_value.dtype not in (torch.uint8,):
```
```suggestion
                if new_value.dtype != torch.uint8:
```

poedator replied:
@younesbelkada ,
thank you for the supportive comments. I will make a major update once the bnb PR goes thru.
meanwhile, could you comment on my message [On storage format for the quantized parameters ](https://github.com/huggingface/transformers/pull/26037#issuecomment-1712438542)above? - I am not very comfortable with flooding the state dict with all components of quant_state as this PR does now.

poedator replied:
Done. `new_value.dtype != torch.uint8` suggestion accepted.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -729,6 +730,16 @@ def _load_state_dict_into_meta_model(
         elif not is_quantized:
             # For backward compatibility with older versions of `accelerate`
             set_module_tensor_to_device(model, param_name, param_device, **set_module_kwargs)
+        elif param.dtype == torch.uint8:
```
for stronger check, you can also add this condition (and just in case some uint8 tensors end up in the state dict of non-quantized models, to be on the safe zone)
```suggestion
        elif param.dtype == torch.uint8 and is_quantized:
```

poedator replied:
Done. `is_quantized` is asserted now and then code branches into 4 / 8 bits.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -3460,14 +3475,18 @@ def _find_mismatched_keys(
-                        )
-                        del state_dict[checkpoint_key]
+                    if model_key in model_state_dict:
+                        if (
+                            state_dict[checkpoint_key].shape[-1] == 1
+                            and state_dict[checkpoint_key].numel() * 2 == model_state_dict[model_key].numel()
+                        ):
+                            # such mismatched weights are OK for 4bit quantizations. TODO: use config-based condition
+                            pass
```
here you can also enforce this condition with `is_quantized`, or even better with `load_in_4bit` to avoid unexpected behaviour with non-quantized models. Also there seems to be a TODO which is a bit unclear from my perspective, could you please elaborate a bit more 🙏 ?

poedator replied:
This `load_in_4bit ` was exactly what I was looking for, however it is not passed to `_find_mismatched_keys` or even `_load_pretrained_model()`. An even better alternative would be to check module type - still requires to bring whole model to this function and search for module. I left it as is, edited the comment for clarity.

younesbelkada reviewed (commented):
Thanks a lot for your great work on this, and for leading this effort! This PR will open a lot of cool perspectives in the near future and help the community by making it possible to push bnb 4-bit weights on the Hub. 
First of all, let's wait for the next bnb release before merging this, that would include https://github.com/TimDettmers/bitsandbytes/pull/753 

Overall the changes look ok to me, I would suggest to enforce the conditions for only 4-bit cases to make sure we don't have any surprising behaviours in case of non-quantized models that stores uint8 buffers. I made few comments, please have a look and let me know what do you think.

I would suggest to add tests for this, similarly as this test: https://github.com/huggingface/transformers/blob/main/tests/quantization/bnb/test_mixed_int8.py#L333 , you could for instance override this test: https://github.com/huggingface/transformers/blob/main/tests/quantization/bnb/test_4bit.py#L193 and write something similar than the 8bit tests

To run the tests, simply run:
```bash
RUN_SLOW=1 pytest tests/quantization/bnb/test_4bit.py
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26037). All of your documentation changes will be reflected on that endpoint.

younesbelkada commented:
Thanks a lot @poedator for your great effort here! Let me know once this is ready for review!

poedator commented:
> Thanks a lot @poedator for your great effort here! Let me know once this is ready for review!

@younesbelkada , thanks for your friendly attention to my work!

I updated the PR code to address your comments and also simplify/clarify some areas. Also stole the PR name from your #22177 ).

For saving format for 4bit params I changed scheme (see bnb my PR for more details) to this:
- save main weight tensor as Parameter using standard state_dict() code
- save tensor components from quant_state into state_dict() with custom code
- pack all other quant_state components into json->bytestring->int8_tensor and save to state dict using name like `...weight.quant_state.bitsandbytes__nf4`
- no extra per-weight metadata stored to safetensors
As a result, we have 6 items stored per weight (with double quant) vs ~12 in earlier draft.

Such scheme can be easily adopted for other quantizers: when loading, `_load_state_dict_into_meta_model()` checks for presence of `[param_name].quant_state.[QUANT_TYPE]` key in loaded dict and use QUANT_TYPE to call appropriate function to such weight in quantized form. It allows to load models with custom choice of quantizers on per-weight basis. Once this PR goes thru, I want to propose how to refactor model_utils to have a uniform interface for quantied models. 

I added tests covering saving / loading and matching by component of tensors and quant_states, + forward() and generate().

Now we need to see the enabling [PR in bnb](https://github.com/TimDettmers/bitsandbytes/pull/753) going thru. Other than that, this PR looks ready for your review.

Yu-Yuqing commented:
Thanks a lot for you work, could you tell me how the 4bit quantized model should be merged with the adapter trained by lora?

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -39,6 +41,9 @@ class `Int8Params` from `bitsandbytes`.
             The value of the tensor (useful when going from the meta device to any other device).
         fp16_statistics (`torch.HalfTensor`, *optional*):
             The list of fp16 statistics to set on the module, used for serialization.
+        quantized_stats dict[str, Any]:
```
```suggestion
        quantized_stats (`dict[str, Any]`, *optional*):
```

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +79,55 @@ class `Int8Params` from `bitsandbytes`.
 
             kwargs = old_value.__dict__
+
+            assert prequantized_loading == (
```
No assert in the codebase please, perform a simple check and raise a propriate error please! 
We avoid this to outcome potential issues, see for instance: https://snyk.io/blog/the-dangers-of-assert-in-python/#

poedator replied:
agree. changed to `raise ValueError`

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +79,55 @@ class `Int8Params` from `bitsandbytes`.
+                is_8bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse(
+                    "0.37.2"
+                )
+                if new_value.dtype in (torch.int8, torch.uint8):
+                    if not is_8bit_serializable:
```
```suggestion
                if new_value.dtype in (torch.int8, torch.uint8) and not is_8bit_serializable:
```

poedator replied:
accepted the edit

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +79,55 @@ class `Int8Params` from `bitsandbytes`.
             module._parameters[tensor_name] = new_value
-            if fp16_statistics is not None:
-                setattr(module.weight, "SCB", fp16_statistics.to(device))
+
+        else:
+            pass  # param.device.type == "cuda"  # is it a possible loading scenario?
```
```suggestion
```
Hmm I don't really understand this 🤔 Can you elaborate more ? I suspect it is not needed so I propose to remove it

poedator replied:
since the preceding branch above is `if param.device.type != "cuda":` I thouhgt that the opposite should be handled too. I changed this code to raise `ValueError`.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -735,16 +736,37 @@ def _load_state_dict_into_meta_model(
-                fp16_statistics = state_dict[param_name.replace("weight", "SCB")]
-            else:
-                fp16_statistics = None
+            assert is_quantized, "something broke above this line"
```
same as above

younesbelkada replied:
Hmm I am also confused about this statement, why would something broke at this point?

poedator replied:
it was done when I was implementing [this suggestion above](https://github.com/huggingface/transformers/pull/26037#discussion_r1321097411). 
I changed `elif` code to be closer to your original suggestion and reworked the lines below it.

poedator replied:
it appeared when I was working on [this comment.](https://github.com/huggingface/transformers/pull/26037#discussion_r1321097411)
changed it closer to your original suggestion + reworked code below this line a bit.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -1829,10 +1851,12 @@ def save_pretrained(
-                "You are calling `save_pretrained` on a 4-bit converted model. This is currently not supported"
-            )
+        if getattr(self, "is_loaded_in_4bit", False) and getattr(self, "is_4bit_serializable", False):
+            warnings.warn(
+                "You are calling `save_pretrained` to a 4-bit converted model you may likely encounter unexepected"
+                " behaviors. If you want to save 4-bit models, make sure to have `bitsandbytes>=0.42` installed.",
+                UserWarning,
+            )  # TODO:  verify version
```
I think we should raise an error here and not allow users to save 4bit models if they have the wrong bnb version - the condition that we have put for 8bit is a bit too flexible

poedator replied:
done. changed to `NotImplementedError` for both 8 and 4 bits.

younesbelkada replied:
Nice, great !

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -3629,8 +3662,12 @@ def _find_mismatched_keys(
 
         if is_quantized:
-            unexpected_keys = [elem for elem in unexpected_keys if "SCB" not in elem]
-            missing_keys = [elem for elem in missing_keys if "SCB" not in elem]
+            # unexpected_keys = [elem for elem in unexpected_keys if "SCB" not in elem]
+            # handled inside _load_state_dict_into_meta_model() now
+
+            # missing_keys = [elem for elem in missing_keys if "SCB" not in elem]
+            # this is not needed -- 8bit model breaks without this item
+            pass
```
If this is handled above let's remove that block

younesbelkada reviewed (commented):
Very nice work and impactful addition @poedator ! Also nice naming of the PR ! 
I left few open questions to better understand your approach, let me know what do you think. 
Can you also confirm 8-bit slow tests pass on your end?

younesbelkada commented:
Also I would like @SunMarc to have a look on the PR whenever possible, as he is also maintaining the quantization feature in transformers, thanks!

poedator commented:
> Thanks a lot for you work, could you tell me how the 4bit quantized model should be merged with the adapter trained by lora?

Hi, @Yu-Yuqing 
I believe, such merge is done by [peft](https://github.com/huggingface/peft) package. It was implemented recently by https://github.com/huggingface/peft/pull/851.
This PR is not affecting such merges. But if(when) it gets merged, it would be possible to save 4-bit quantized models, including ones that had adapters merged into them.

poedator commented:
[UPDATE ][WiP] 
- the slow tests are a bit broken, give me some time to fix it before continuing review
- meanwhile at bnb we had  first round of comments.

poedator commented:
Update: I am still optimistic that the enabling PR in BitsAndBytes will get merged soon, but can't bet on a specific date. 

As an alternative, I can try to bring 4-bit serialization code to here. But that would require subclassing `bnb.Linear4bit` with risk of compatibility issues.  Does it make sense?  @younesbelkada

younesbelkada commented:
Thanks a lot @poedator ! 
I think it is better to have the bnb serialization live directly in bnb, there is no rush, we can wait until your PR gets merged!

poedator commented:
rebased to main as of Nov 8. 
Created test model https://huggingface.co/poedator/opt-125m-bnb-4bit using this PR and [latest relevant PR in bnb](https://github.com/TimDettmers/bitsandbytes/pull/864/files). Presumably it is ready for testing once the bnb PR merges.

younesbelkada reviewed (commented):
Hi @poedator 
Thanks very much for leading the effort of making bnb 4bit models even more accessible! 🤩 
I played a bit with your PR and I am facing a strange issue which might be related to bnb when I use ` bnb_4bit_use_double_quant=False`. I think that the tests you have designed could be much more simpler and simply just check if the model loaded with `load_in_4bit` and `from_pretrained` leads to the same `get_memory_footprint()` as well as same generation quality, for 4 major scenarios:

Scenario 1:
```
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=False,
    bnb_4bit_compute_dtype=torch.bfloat16,
)
```

Scenario 2:
```
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.bfloat16,
)
```

Scenario 3:
```
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="fp4",
    bnb_4bit_use_double_quant=False,
    bnb_4bit_compute_dtype=torch.bfloat16,
)
```

Scenario 4:
```
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="fp4",
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.bfloat16,
)
```

Reproducible snippet (that you can use for the tests):

```python
import tempfile
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("bigscience/bloom-560m", load_in_4bit=True)
tok = AutoTokenizer.from_pretrained("bigscience/bloom-560m")

prompt = "Write a function that takes a list of numbers and returns the sum of the list."

input_text = f"""
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
{prompt}

### Response:
"""

inputs = tok(input_text, return_tensors="pt").to(0)
outputs = model.generate(inputs.input_ids, max_new_tokens=100)

print(tok.decode(outputs[0]))

with tempfile.TemporaryDirectory() as tmp_dir:
    model.save_pretrained(tmp_dir, safe_serialization=False)
    model = AutoModelForCausalLM.from_pretrained(tmp_dir)
    out = model.generate(input_ids=inputs.input_ids)
```

Error trace:

```bash
  File "/home/younes_huggingface_co/code/transformers/src/transformers/integrations/bitsandbytes.py", line 121, in set_module_quantized_tensor_to_device
    new_value = bnb.nn.Params4bit.from_prequantized(
  File "/home/younes_huggingface_co/miniconda3/envs/fix-test/lib/python3.9/site-packages/bitsandbytes/nn/modules.py", line 161, in from_prequantized
    self.quant_state = QuantState.from_dict(qs_dict=quantized_stats, device=device)
  File "/home/younes_huggingface_co/miniconda3/envs/fix-test/lib/python3.9/site-packages/bitsandbytes/functional.py", line 645, in from_dict
    shape=torch.Size(qs_dict['shape']),
TypeError: 'NoneType' object is not iterable
```

Thanks again for all your efforts!

poedator commented:
@younesbelkada , thank you for  picking this up so fast. Indeed, `bnb_4bit_use_double_quant=False` causes error in serialization, it will be fixed on the bnb side. 
Meanwhile, let me know if any other issues pop up when testing with bnb_4bit_use_double_quant=True.
Also I need advice on shared tensors and on pytorch test - see my previous message.

younesbelkada commented:
Thanks @poedator ! 
For the safetensors issue please refer to this PR: https://github.com/huggingface/transformers/pull/27234 I had to fix something similar for llm.int8 weights. Let me know if this PR helps you digging a bit!

poedator commented:
> Contributor

the example shows how to identify shared pointers. But where is that collection stored then and how it is used in unpacking?

KnutJaegersberg commented:
Yesterday Tim tweeted one can use 4 bit save with an bitsandbytes update. I tried by pip installing the pull requested transformers version from source. I could save the model, but on loading it, I got this exception: 

>>> model=AutoModel.from_pretrained("/run/media/knut/HD2/Deacon-34b-4bit/")
Loading checkpoint shards:   0%|                                                             | 0/4 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/knut/Downloads/transformers-save4/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    return model_class.from_pretrained(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3532, in from_pretrained
    ) = cls._load_pretrained_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3927, in _load_pretrained_model
    new_error_msgs, offload_index, state_dict_index = _load_state_dict_into_meta_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 761, in _load_state_dict_into_meta_model
    recover_shared_tensors_values(state_dict, quantized_stats, param_name + ".", likely_shared_key_endings)
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 799, in recover_shared_tensors_values
    target_dict[target_prefix + ending] = replacement_tensors[0]
IndexError: list index out of range

KnutJaegersberg commented:
I got the same issue for Mistral, too.

poedator commented:
> I got the same issue for Mistral, too.

hi, @KnutJaegersberg , thank you for the interest to bnb 4bit serialization. While most of the changes are already made on bnb side, for successful 4bit serialization one may need to wait for this PR to merge. Hopefully it is a matter of days.

KnutJaegersberg commented:
Thanks for your answer, @poedator I started thinking I just made a stupid mistake. I'll try what you said as soon as my local compute is available again.

KnutJaegersberg commented:
Same exception with save_serialization=False. 
hmm
Loading checkpoint shards:   0%|                                                                                                                          | 0/4 [00:01<?, ?it/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/knut/Downloads/transformers-save4/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    return model_class.from_pretrained(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3532, in from_pretrained
    ) = cls._load_pretrained_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3927, in _load_pretrained_model
    new_error_msgs, offload_index, state_dict_index = _load_state_dict_into_meta_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 761, in _load_state_dict_into_meta_model
    recover_shared_tensors_values(state_dict, quantized_stats, param_name + ".", likely_shared_key_endings)
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 799, in recover_shared_tensors_values
    target_dict[target_prefix + ending] = replacement_tensors[0]
IndexError: list index out of range

younesbelkada commented:
Hi @poedator 

> the example shows how to identify shared pointers. But where is that collection stored then and how it is used in unpacking?

It is used directly by storing them on the state dict, that is what we do for 8bit layers.

poedator commented:
@younesbelkada 
I decided to un-share these nested_quant_map tensors, which are only 256 elements each. So there is no need to deal with restoring them for 4bit quantization. 

I completed the needed changes on bnb side ([PR pending](https://github.com/TimDettmers/bitsandbytes/pull/868)). Also added a set of 4-bit serialization tests in a separate file. They are more detailed than you suggested, but this gives me mode peace of mind. 

Please resume the review.
---------------
@KnutJaegersberg - you may want to re-test this PR now. That problem should be gone.

KnutJaegersberg commented:
@poedator thank you :) 
I tried again after reinstalling tf, now I get another exception: 

Loading checkpoint shards:   0%|                                                                                                                          | 0/4 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/knut/Downloads/transformers-save4/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    return model_class.from_pretrained(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3512, in from_pretrained
    ) = cls._load_pretrained_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3907, in _load_pretrained_model
    new_error_msgs, offload_index, state_dict_index = _load_state_dict_into_meta_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 760, in _load_state_dict_into_meta_model
    set_module_quantized_tensor_to_device(
  File "/home/knut/Downloads/transformers-save4/src/transformers/integrations/bitsandbytes.py", line 121, in set_module_quantized_tensor_to_device
    new_value = bnb.nn.Params4bit.from_prequantized(
  File "/home/knut/transformers/lib/python3.9/site-packages/bitsandbytes/nn/modules.py", line 161, in from_prequantized
    self.quant_state = QuantState.from_dict(qs_dict=quantized_stats, device=device)
  File "/home/knut/transformers/lib/python3.9/site-packages/bitsandbytes/functional.py", line 645, in from_dict
    shape=torch.Size(qs_dict['shape']),
TypeError: 'NoneType' object is not iterable

poedator commented:
@KnutJaegersberg ,
I am not seeing this error in my tests, but to be on the safe side, I edited the line 645 in the BNB PR.
Please help me understand what triggered it? Could it be that you try to load a model created with earlier version of either PRs? Try to run the whole cycle "quantize-save-load" using the latest PR commits. Or try loading [my model form HF](https://huggingface.co/poedator/opt-125m-bnb-4bit).

KnutJaegersberg commented:
@poedator I I did not use the latest PR of BNB because it gave me a cuda setup exception. It might be a local issue then. Thanks!

KnutJaegersberg commented:
I'm trying to do something weird, I guess. I can load your opt model @poedator but not my llamafied yi-34b, though I tried it with the same config, double quantization, no safetensors. 

I'll upload my weights here, it might take a few hours:
https://huggingface.co/KnutJaegersberg/Deacon-34b-4bit


  File "/home/knut/Downloads/transformers-save4/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    return model_class.from_pretrained(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3512, in from_pretrained
    ) = cls._load_pretrained_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 3907, in _load_pretrained_model
    new_error_msgs, offload_index, state_dict_index = _load_state_dict_into_meta_model(
  File "/home/knut/Downloads/transformers-save4/src/transformers/modeling_utils.py", line 760, in _load_state_dict_into_meta_model
    set_module_quantized_tensor_to_device(
  File "/home/knut/Downloads/transformers-save4/src/transformers/integrations/bitsandbytes.py", line 121, in set_module_quantized_tensor_to_device
    new_value = bnb.nn.Params4bit.from_prequantized(
  File "/home/knut/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/nn/modules.py", line 161, in from_prequantized
    self.quant_state = QuantState.from_dict(qs_dict=quantized_stats, device=device)
  File "/home/knut/miniconda3/envs/textgen/lib/python3.10/site-packages/bitsandbytes/functional.py", line 633, in from_dict
    code=qs_dict['nested_quant_map'].to(device),
KeyError: 'nested_quant_map'

poedator commented:
@KnutJaegersberg , 
Thank you for your persistence! I was able to reproduce the error with `llamafied yi-34b`.  Will post my findings and fixes here.

KnutJaegersberg commented:
@poedator I had issues building bnb for your newest commit on my local environment, have not resolved those so far. I tried to manually do the changes of your bnb commit to my local files, my transformers library should be exactly the same. It's properly my environment which is the cause. 
I'll just wait till the next bnb release, thanks :)

SunMarc commented:
cc @Titus-von-Koeller

estibi commented:
Hello, any update on this issue? Or is there any other way to save merged 4 bit model?

poedator committed:
updated bitsandbytes.py

poedator committed:
rm test_raise_* from test_4bit.py

poedator committed:
add test_4bit_serialization.py

poedator committed:
modeling_utils bulk edits

poedator commented:
With release of `bitsandbytes==0.41.3` the path to this PR seems clear. (Big thanks to @Titus-von-Koeller !)
I rebased the code to the latest main and retested.
All relevant long tests in quantization section pass.
@younesbelkada @SunMarc - please review.

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +77,56 @@ class `Int8Params` from `bitsandbytes`.
-                new_value = bnb.nn.Params4bit(new_value, requires_grad=False, **kwargs).to(device)
-
+                if prequantized_loading:
+                    is_4bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse(
+                        "0.41"
```
```suggestion
                    is_4bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) >= version.parse(
                        "0.41.3"
```

younesbelkada commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +77,56 @@ class `Int8Params` from `bitsandbytes`.
+                    is_4bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse(
+                        "0.41"
+                    )
+                    # TODO update version number after BNB release with PR #753
```
```suggestion
```

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2160,18 +2184,23 @@ def save_pretrained(
+        if (
+            getattr(self, "is_loaded_in_4bit", False)
+            and getattr(self, "is_4bit_serializable", False)
+            and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.41.0")
```
```suggestion
            and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.41.3")
```

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -4119,10 +4158,6 @@ def _find_mismatched_keys(
                 )
             raise RuntimeError(f"Error(s) in loading state_dict for {model.__class__.__name__}:\n\t{error_msg}")
 
-        if is_quantized:
```
Why this has been removed?

poedator replied:
this part of code was originally meant to deal with loaded components that are not params or buffers. Now this is addressed in  `load_state_dict_into_meta_model()` [here](https://github.com/huggingface/transformers/pull/26037/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR774)

poedator replied:
I thought that having in the main function code for dealing with individual components (and adding 4bit stuff there) was not pretty.

In #26610 I take this idea further and [have a method update_mismatched_keys()](https://github.com/huggingface/transformers/pull/26610/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR3620) for the same purpose.

younesbelkada commented on tests/quantization/bnb/test_4bit_serialization.py:
```diff
@@ -0,0 +1,189 @@
+@require_accelerate
+@require_torch
+@require_torch_gpu
+# @slow
```
```suggestion
@slow
```

younesbelkada commented on tests/quantization/bnb/test_4bit_serialization.py:
```diff
@@ -0,0 +1,189 @@
+@require_torch
+@require_torch_gpu
+# @slow
+class BaseTest(unittest.TestCase):
```
Can you move this test inside `tes4bit.py`?

poedator replied:
done

younesbelkada reviewed (commented):
Thanks a lot @poedator for your work on this - I left a couple of comments regarding the bnb minimal version and about tests, can you please have a look? 🙏 Thanks!

poedator committed:
bnb_ver 0.41.3 in integrations/bitsandbytes.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
@slow reinstated

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
bnb ver 0.41.3 in  src/transformers/modeling_utils.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
rm bnb version todo in  integrations/bitsandbytes.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
moved 4b serialization tests to test_4bit

poedator committed:
tests upd for opt

poedator commented:
> Thanks a lot @poedator for your work on this - I left a couple of comments regarding the bnb minimal version and about tests, can you please have a look? 🙏 Thanks!

@younesbelkada , thank you for pointing out the things that I overlooked.
I left one comment and fixed the rest. Pls check if it can fly now.

Titus-von-Koeller commented:
Hey all, progress on this PR is looking good. Thanks for all your work!

Just wanted to quickly make you aware, that due to failing PEFT tests, we had to [revert this change to Params4bit](https://github.com/TimDettmers/bitsandbytes/commit/f2e4b6d01f243f08c191bf6aaceb7ba281337354) and [just released a postfix to PyPi](https://pypi.org/project/bitsandbytes/0.41.3.post1/).

Not sure if this affects the PR here, but just wanted to be sure this final review takes this most recent change into account.

poedator commented:
> Hey all, progress on this PR is looking good. Thanks for all your work!
> 
> Just wanted to quickly make you aware, that due to failing PEFT tests, we had to [revert this change to Params4bit](https://github.com/TimDettmers/bitsandbytes/commit/f2e4b6d01f243f08c191bf6aaceb7ba281337354) and [just released a postfix to PyPi](https://pypi.org/project/bitsandbytes/0.41.3.post1/).
> 
> Not sure if this affects the PR here, but just wanted to be sure this final review takes this most recent change into account.

@Titus-von-Koeller thanks for the notice. I tried to re-test with 0.41.3.post1, [but got RTE, details here](https://github.com/TimDettmers/bitsandbytes/issues/907) 0.41.3. was OK.

younesbelkada reviewed (commented):
Awesome work ! Left a couple of nits, can you confirm 4bit and 8bit slow tests pass with these changes?

poedator committed:
to torch_device

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator commented on tests/quantization/bnb/test_4bit.py:
```diff
@@ -521,3 +510,138 @@ def test_training(self):
+    input_text = "Mars colonists' favorite meals are"
+
+    def setUp(self):
+        self.device = torch.device("cuda:0")
```
done. BTW, `torch_device` had to be imported explicitly from `transformers.testing_utils`

poedator commented:
> Awesome work ! Left a couple of nits, can you confirm 4bit and 8bit slow tests pass with these changes?

something is not right between this PR and accelerate now. I get `ValueError: weight is on the meta device, we need a `value` to put in on cuda.`
Give me some time to debug. 
Besides, let's wait for this issue to resolve at bnb: https://github.com/TimDettmers/bitsandbytes/issues/907

Titus-von-Koeller commented:
We managed to fix the issue @poedator had with the bnb install. My GCC was too new and produces binaries that are not backward compatible enough. We now reissued the last release to be more backward compatible and could confirm that it works with an older GCC dependency.

The cuda library is packaged with a c++ standard library. If the user does not fulfill at least the c++ standard of the linked library, it cannot run and throws an error.

younesbelkada commented:
Awesome work, thanks everyone ! So this PR is ready for final review?

poedator committed:
ruff fixes to tests

poedator commented:
Well, for a moment I thought that we are done, prepared to see this PR merge and celebrate. But then in the last minute the `tests/quantization/bnb/test_4bit.py::ExtendedSerializationTest` tests started failing with odd error `ValueError: weight is on the meta device, we need a value to put in on cuda`. This was especially odd, since almost same code worked flawlessly in my own tests which looked very similar. 

During debug, I found that the initially loaded model's weights were missing `quant_state` components, and therefore they were not saved and not available for subsequent loading. 
The `quant_state` components were created OK in `_load_state_dict_into_meta_model()` but later stripped out in `dispatch_model()`. With deeper dive into accelerate I found that magic happens in `accelerate.modeling.py::set_module_tensor_to_device()`:

1) [in this line the code took wrong branch](https://github.com/huggingface/accelerate/blob/ad3a5bc92084b9c5c6d5e141eb5ff0029f3fcc78/src/accelerate/utils/modeling.py#L327) because apparently `device(type='cuda') != device(type='cuda', index=0)`
`torch.device(device) != module._parameters[tensor_name].device`
`torch.device(device)` comes from device_map, based on `torch_device`, equal to `"cuda"` string
`module._parameters[tensor_name].device` is the result of torch `.cuda()` call in `bnb.Params4bit.cuda()`.
note that `torch.tensor([42]).to('cuda').device` returns `device(type='cuda', index=0)`

2) once in the wrong branch, the [code gets to line](https://github.com/huggingface/accelerate/blob/ad3a5bc92084b9c5c6d5e141eb5ff0029f3fcc78/src/accelerate/utils/modeling.py#L342) `new_value = param_cls(new_value, requires_grad=old_value.requires_grad)` and happily creates a `Param4bit` parameter with just 4bit weights and none of the `quant_state` components, which causes error much later.

3) Also found an issue with class name https://github.com/huggingface/accelerate/issues/2250 which is not relevant to this problem

It was not apparent until we swithed from `"cuda:0"` to `"cuda"` in this commit https://github.com/huggingface/transformers/pull/26037/commits/6834f084e488737bc7252ffff093eb2a470258b1

Ideas for how to deal with this (pick any number of them):
1. compare not devices, but device types in [this accelerate line](https://github.com/huggingface/accelerate/blob/ad3a5bc92084b9c5c6d5e141eb5ff0029f3fcc78/src/accelerate/utils/modeling.py#L327) 
1. convert `device(type='cuda')` into `device(type='cuda', index=0)` along the way, for instance in `accelerate::dispatch_model()` 
2. simplify overal quantized weights handling, avoiding additional tinkering in accelerate, i.e move all logic from `set_module_tensor_to_device()` to `transformers`.
1. make bnb.Param4bit constructor more robust, so it would test data at init phase (and fail in case it gets just the qunatized weights like above).
1. asking `pytorch` to make `device(type='cuda')` equal `device(type='cuda', index=0)`  (makes sense but takes time and may cause other issues)
1. in transformers testing or consider testing with both `"cuda"` and `device(type='cuda', index=0)`, and
1. re-test the rest of HF code to see where `(device(type='cuda') != device(type='cuda', index=0)` problem may arise.

@younesbelkada , need your guidance and/or changes in `accelerate`.
cc @SunMarc @Titus-von-Koeller

younesbelkada commented:
Hi @poedator 
Again thank you very much for all your efforts to push this forward! 
Thanks also for providing a detailed description of the new problem. The fix IMO should go into accelerate: https://github.com/huggingface/accelerate/pull/2254 
I suggest to wait until we get feedback from accelerate maintainers and see if this approach gets validated, if that's the case,  we'll merge this PR together with the accelerate PR and tell users to install accelerate from source if they face into a similar issue. Can you also meanwhile confirm https://github.com/huggingface/accelerate/pull/2254 fixes the problem?

SunMarc commented:
The PR is merged ! LMK @poedator if this fixes the issue

poedator commented:
@younesbelkada, @SunMarc ! thank you for the quick reaction. I re-tested with https://github.com/huggingface/accelerate/pull/2254 and the problem is gone now. Hope that the tests are all green on your side too, looking forward to the merge.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2160,18 +2184,23 @@ def save_pretrained(
-            and not getattr(self, "is_8bit_serializable", False)
-            and not _hf_peft_config_loaded
+            and getattr(self, "is_8bit_serializable", False)
+            and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.37.2")
```
```suggestion
```
This is already checked by `is_8bit_serializable` attribute no?

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2160,18 +2184,23 @@ def save_pretrained(
-            and not _hf_peft_config_loaded
+            and getattr(self, "is_8bit_serializable", False)
+            and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.37.2")
         ):
```
```suggestion
            and not _hf_peft_config_loaded
        ):
```
Why this has been removed?

poedator replied:
agree. perhaps lost it during another rebase

younesbelkada reviewed (commented):
Thanks a lot for iterating @poedator ! I left three final nits - looking forward to merge this PR !

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2160,18 +2184,23 @@ def save_pretrained(
+        if (
+            getattr(self, "is_loaded_in_4bit", False)
+            and getattr(self, "is_4bit_serializable", False)
+            and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.41.3")
```
and same as above, let's remove the. `and version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.41.3")` check

poedator committed:
rm redundant bnb version check in mod_utils

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
restore _hf_peft_config_loaded  modeling_utils.py::2188

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
restore _hf_peft_config_loaded  test in modeling_utils.py::2199

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2160,18 +2184,23 @@ def save_pretrained(
         if (
             getattr(self, "is_loaded_in_8bit", False)
-            and not getattr(self, "is_8bit_serializable", False)
+            and getattr(self, "is_8bit_serializable", False)
```
```suggestion
            and not getattr(self, "is_8bit_serializable", False)
```

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2740,8 +2769,10 @@ def from_pretrained(
+            is_4bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) >= version.parse("0.41.3")
             is_8bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse("0.37.2")
         else:
+            is_4bit_serializable = False
```
It seems `is_4bit_serializable` is not set into the model - can you set it here: https://github.com/poedator/transformers/blob/save4/src/transformers/modeling_utils.py#L3510 ?

younesbelkada replied:
Or is it somehow silently set somewhere else?

poedator replied:
[is_4bit_serializable is set just above, in 2772.](https://github.com/huggingface/transformers/pull/26037/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR2772) Or did you mean something else?

younesbelkada replied:
I meant setting it as an attribute into the model: `model.is_4bit_serializable = is_4bit_serializable`

younesbelkada reviewed (commented):
Thanks a lot for iterating!
Is `is_4bit_serializable` set into the model somewhere in `from_pretrained` ? Inspecting your branch I can't find it

poedator committed:
fixed NOT getattr(self, "is_8bit_serializable")

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

poedator committed:
setting model.is_4bit_serializable

poedator commented:
> Thanks a lot for iterating! Is `is_4bit_serializable` set into the model somewhere in `from_pretrained` ? Inspecting your branch I can't find it
you were correct about that. Fixed. Slow tests all pass. Pls review.

younesbelkada reviewed (approved):
Thanks a lot for this @poedator ! 
I will let @SunMarc and @amyeroberts give their review on this PR before we merge

SunMarc commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +77,55 @@ class `Int8Params` from `bitsandbytes`.
+                is_8bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) > version.parse(
+                    "0.37.2"
+                )
+                if new_value.dtype in (torch.int8, torch.uint8) and not is_8bit_serializable:
```
Either simplify this condition just like the 4-bit case or add the same condition for 4-bit. This depends on if we only use `set_module_quantized_tensor_to_device` when we have a quantized value.

poedator replied:
added similar check on 4-bit side to be safe

SunMarc commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +77,55 @@ class `Int8Params` from `bitsandbytes`.
-            if fp16_statistics is not None:
-                setattr(module.weight, "SCB", fp16_statistics.to(device))
+        else:
+            raise ValueError("Quantized parameter passed with device.type == 'cuda'")
```
Have you experienced this specific case ? Maybe we can explain a little bit why (param already on cuda, we should not move it again)

poedator replied:
this `else` was set to complete [the `if` from original code](https://github.com/huggingface/transformers/blob/ebfdb9ca62205279d5019ef1403877461b3b2da4/src/transformers/integrations/bitsandbytes.py#L72). In the original code the else branch was absent and nothing happened when param was already on cuda. It seemed counterintuitive, so I added the `else` branch with raising error. I do not remember it triggered though.

I will delete this else statement to bring the code closer to the original.

This original `if` looks redundant to me in standard cases where the model is loaded to `meta` device first. I hesitate to delete it entirely myself though.

SunMarc commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -39,6 +41,8 @@ class `Int8Params` from `bitsandbytes`.
             The value of the tensor (useful when going from the meta device to any other device).
         fp16_statistics (`torch.HalfTensor`, *optional*):
             The list of fp16 statistics to set on the module, used for serialization.
```
```suggestion
            The list of fp16 statistics to set on the module, used for 8-bit serialization.
```

poedator replied:
done with combinig `fp16_statistics` and `quantized_stats`

SunMarc commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -39,6 +41,8 @@ class `Int8Params` from `bitsandbytes`.
             The value of the tensor (useful when going from the meta device to any other device).
         fp16_statistics (`torch.HalfTensor`, *optional*):
             The list of fp16 statistics to set on the module, used for serialization.
+        quantized_stats (`dict[str, Any]`, *optional*):
+            Dict with items for 4-bit quantization
```
```suggestion
        quantized_stats (`dict[str, Any]`, *optional*):
            Dict with items for 4-bit serialization
```

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -747,17 +748,40 @@ def _load_state_dict_into_meta_model(
+                fp16_statistics_key = param_name.replace("weight", "SCB")
+                unexpected_keys.remove(fp16_statistics_key)
+
+                if "SCB" not in param_name:
+                    # looks like a redundant if -- .SCB should not be in the loaded_state_dict_keys or expected_keys
```
Have you launch some tests to try without that ? If it i redundant, it would be great to remove it.

poedator replied:
tested, observed no harm, removed it

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -747,17 +748,40 @@ def _load_state_dict_into_meta_model(
+                    )
+        else:
+            # loading not quantized params in quantized model
+            set_module_quantized_tensor_to_device(model, param_name, param_device, value=param)
```
can we use simply `set_module_tensor_to_device` ?

poedator replied:
I tried it and got errors in testing. there is something there that messes up quantized bnb parameters, one could be [bug with class name](https://github.com/huggingface/accelerate/issues/2250), but there is more, since 8bit fails too.  In my view, `set_module_tensor_to_device()` should be simplified (with all quantization stuff removed), and brought to  transformers. Or the quantized objects should know how to load themselves properly. I tried to improve that in #26610.

SunMarc replied:
Awesome, them let's try to fix this in the other big PR =)

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -3970,10 +4000,18 @@ def _find_mismatched_keys(
-                            (checkpoint_key, state_dict[checkpoint_key].shape, model_state_dict[model_key].shape)
-                        )
-                        del state_dict[checkpoint_key]
+                        if (
+                            state_dict[checkpoint_key].shape[-1] == 1
+                            and state_dict[checkpoint_key].numel() * 2 == model_state_dict[model_key].numel()
+                        ):
+                            # Such mismatched weights are OK for 4bit quantizations.
+                            # Need more reliable condition here, ideally based on type(module)
```
I don't know how reliable it is

poedator replied:
I thought about it but could not find a better way to verify 4bit param type:
- param object type is not available inside `_find_mismatched_keys()`
- `checkpoint_key` name is generic and it gives no clue whether it is 4-bit param.
- 8bit quant has no such problem because there are 8-bit torch types (unlike 4bit so far)
- using `ignore_mismatched_sizes` seems excessive for this case. 

could you suggest something?

SunMarc replied:
That's fine then! Could you just add a quick note on how these conditions help us to find the 4-bit param type, especially the second condition.

poedator replied:
edited the comment to explain reasons for the exception.
rebased to fresh main.

SunMarc commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -21,7 +21,9 @@
 
-def set_module_quantized_tensor_to_device(module, tensor_name, device, value=None, fp16_statistics=None):
+def set_module_quantized_tensor_to_device(
+    module, tensor_name, device, value=None, fp16_statistics=None, quantized_stats=None
```
The best solution would have been to only have one arg to store the quantized stats for 8-bit or 4-bit but i'm fine having both args as it is clearer.

poedator replied:
agreed and done

SunMarc reviewed (approved):
Thanks for working on this @poedator for such a long time ! This is a very exciting feature that will enable many applications ! Left a few comments but nothing blocking.

poedator commented:
> Thanks for working on this @poedator for such a long time ! This is a very exciting feature that will enable many applications ! Left a few comments but nothing blocking.

Hi, @SunMarc thank you for the comment. They all look relevant and implementable. Let me think how to best address them.

poedator committed:
rm separate fp16_statistics arg from set_module...

poedator committed:
rm else branch in integrations::bnb::set_module

poedator committed:
bnb 4bit dtype check

SunMarc reviewed (approved):
Thanks for iterating ! LGTM ! Can you have a final look @amyeroberts ?

poedator committed:
upd comment on 4bit weights

amyeroberts commented on tests/quantization/bnb/test_4bit.py:
```diff
@@ -521,3 +511,135 @@ def test_training(self):
+    def test_nf4_double_safe(self):
+        self.test_serialization(quant_type="nf4", double_quant=True, safe_serialization=True)
+
+    def test_fp4_single_unsafe(self):
+        self.test_serialization(quant_type="fp4", double_quant=False, safe_serialization=False)
+
+    def test_fp4_double_unsafe(self):
+        self.test_serialization(quant_type="fp4", double_quant=True, safe_serialization=False)
```
Is it possible to have "fp4" and safe serialization?

poedator replied:
good catch! it was my omission; hopefully harmless. I update the tests accordingly.

amyeroberts reviewed (commented):
Beautiful piece of work - thanks for adding! 

All my comments are just questions to help me understand part of the quantization logic. Once answered should be good for approval :)

poedator commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +73,53 @@ class `Int8Params` from `bitsandbytes`.
+                    is_4bit_serializable = version.parse(importlib.metadata.version("bitsandbytes")) >= version.parse(
+                        "0.41.3"
+                    )
+                    if new_value.dtype in (torch.int8, torch.uint8) and not is_4bit_serializable:
```
Correct!  Those parameters are stored in 8-bit dtypes, 2 per byte, to save the bits. That's because there are no native 4bit dtypes in torch (yet).

poedator commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +73,53 @@ class `Int8Params` from `bitsandbytes`.
             elif is_4bit:
-                new_value = bnb.nn.Params4bit(new_value, requires_grad=False, **kwargs).to(device)
-
+                if prequantized_loading:
```
the  check for `prequantized_loading` exists in 8 bit too - see line 101. it is just much less prominent because the 8bit param requires only a single `setattr` to deliver the SCB component to it. 

When working on 4bit serialization I found that it required putting in place several additional components per parameter, so I created a separate method `.from_prequantized() to contain all this magic`.

poedator commented on src/transformers/integrations/bitsandbytes.py:
```diff
@@ -74,32 +73,53 @@ class `Int8Params` from `bitsandbytes`.
+                            "Detected 4-bit weights but the version of bitsandbytes is not compatible with 4-bit serialization. "
+                            "Make sure to download the latest `bitsandbytes` version. `pip install --upgrade bitsandbytes`."
+                        )
+                    new_value = bnb.nn.Params4bit.from_prequantized(
```
in 8bit parameter all is needed at loading stage is to move this SCB component to cuda device and set it in place with `setattr`. So it could be done in 1 line.

the 4bit code requires initialization of internal class `QuantState` and in its `.from_dict()` method(https://github.com/TimDettmers/bitsandbytes/blob/f63abb5a0d0bc971d28972ba890a9e59596caac4/bitsandbytes/functional.py#L603) the quantization components are repacked and put in place on the device. It surely may be streamlined, but I tried not to alter the `bnb` classes logic.
cc @Titus-von-Koeller

poedator committed:
upd tests for FP4 safe

amyeroberts reviewed (approved):
Great piece of work! 

Thanks for taking the time to reply to all of my questions and adding the extra tests

younesbelkada reviewed (approved):
Thanks a lot ! Ran again the 4-bit slow tests and they all pass - thanks for your great work!

RonanKMcGovern commented:
Great work @poedator .

Where will docs be available to read about these updates?

poedator commented:
> Where will docs be available to read about these updates?

@RonanKMcGovern, thanks for the interest to this feature. It should work with standard transformers' `.save_pretrained()` and `.from_pretrained()` methods. Check out [how it is done in the tests](https://github.com/huggingface/transformers/blob/f9a98c476c3a92beaba3b20e51f1ff49417231a6/tests/quantization/bnb/test_4bit.py#L521) or how 8bit quantized models did that.
Meanwhile  I am checking with the maintainers about needs to update documentation if any.

hjenryin commented:
Hi @poedator, thank you for your wonderful work!

If I understand this correctly, this should work with all 4-bit models right? However if I run this directly,
```
from transformers import AutoModelForCausalLM 

model=AutoModelForCausalLM.from_pretrained("facebook/opt-125m",load_in_4bit=True)
model.save_pretrained("opt_cache")
```
The error
```
NotImplementedError: You are calling `save_pretrained` on a 4-bit converted model. This is currently not supported
```
will be raised.

I'm not sure whether this `load_in_4bit` makes any difference and we need to load the 4bit model in another fashion, or is it just because this `NotImplementeError` is not removed?

I'm using `bitsandbytes 0.42.0` and `transformers 4.36.2`, which are already the latest. Thanks a lot!

amyeroberts commented:
Hi @hjenryin, 4.36.2 is the latest version release on pypi, but it doesn't contain all of the most recent commits (including this one). If you want to use the bleeding edge (not guaranteed stable) version of transformers, you can [install from source](https://huggingface.co/docs/transformers/installation#install-from-source). 

Alternatively, you can wait until v4.37 is released, which will include this feature. We release a new version roughly one every month, and are due to release one at the end of this week.

hkhan9 commented:
> Hi @hjenryin, 4.36.2 is the latest version release on pypi, but it doesn't contain all of the most recent commits (including this one). If you want to use the bleeding edge (not guaranteed stable) version of transformers, you can [install from source](https://huggingface.co/docs/transformers/installation#install-from-source).
> 
> Alternatively, you can wait until v4.37 is released, which will include this feature. We release a new version roughly once every month, and are due to release one at the end of this week.

Hi @amyeroberts, thanks for putting this. I am having the same issue. I am curious when v4.37 would be published with these changes. Please let me know if you have any estimated timeline? Thanks for all the work.

amyeroberts commented:
Hi @hkhan9, we typically have new releases every month. The release for v4.37 is a bit later than that because of the Christmas and new year period. It will likely be the end of this week or the next.

hkhan9 commented:
> Hi @hkhan9, we typically have new releases every month. The release for v4.37 is a bit later than that because of the Christmas and new year period. It will likely be the end of this week or the next.

Thank you @amyeroberts

Titus-von-Koeller commented:
Thanks @poedator for this very helpful contribution and your (everyone
elses) persistence in getting it across the finishing line! I think this
was a significant step in a good direction.

On Wed, Jan 17, 2024, 16:05 hkhan9 ***@***.***> wrote:

> Hi @hkhan9 <https://github.com/hkhan9>, we typically have new releases
> every month. The release for v4.37 is a bit later than that because of the
> Christmas and new year period. It will likely be the end of this week or
> the next.
>
> Thank you @amyeroberts <https://github.com/amyeroberts>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/26037#issuecomment-1896000621>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFBEO2YJMG6YTGLCARJMETYO7SC5AVCNFSM6AAAAAA4PLZTBGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJWGAYDANRSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

webpolis commented:
Hi

I first uninstalled **transformers** and **bitsandbytes*. I then install transformers from source, hoping it'll include the proper bnb version:

```sh
pip install git+https://github.com/huggingface/transformers.git
```

But, when I load a 4bits model I previously saved using **poedator**'s fork, which uses to work perfectly using that fork:

```python
        model = AutoModelForCausalLM.from_pretrained(
            model_name_or_path,
            low_cpu_mem_usage=True,
            local_files_only=model_name_or_path[0] == '.' or model_name_or_path[0] == '/',
            device_map=device_map if device_map else None,
            # torch_dtype=torch.float32
        )
```

It fails with following error:

```python
File ~/anaconda3/envs/musai/lib/python3.11/site-packages/transformers/integrations/bitsandbytes.py:58, in set_module_q
uantized_tensor_to_device(module, tensor_name, device, value, quantized_stats)
     55 is_buffer = tensor_name in module._buffers
     56 old_value = getattr(module, tensor_name)
---> 58 if old_value.device == torch.device("meta") and device not in ["meta", torch.device("meta")] and value is None
:
     59     raise ValueError(f"{tensor_name} is on the meta device, we need a `value` to put in on {device}.")
     61 prequantized_loading = quantized_stats is not None

NameError: name 'torch' is not defined
```

Then, I restart the environment and install latest **bitsandbytes** from source:

```sh
pip install git+https://github.com/TimDettmers/bitsandbytes.git
```

But, when loading the model again (in a fresh env), it fails:

```python
File ~/anaconda3/envs/musai/lib/python3.11/site-packages/transformers/integrations/bitsandbytes.py:113, in set_module_
quantized_tensor_to_device(module, tensor_name, device, value, quantized_stats)
    108     if new_value.dtype in (torch.int8, torch.uint8) and not is_4bit_serializable:
    109         raise ValueError(
    110             "Detected 4-bit weights but the version of bitsandbytes is not compatible with 4-bit serialization
. "
    111             "Make sure to download the latest `bitsandbytes` version. `pip install --upgrade bitsandbytes`."
    112         )
--> 113     new_value = bnb.nn.Params4bit.from_prequantized(
    114         data=new_value,
    115         quantized_stats=quantized_stats,
    116         requires_grad=False,
    117         device=device,
    118         **kwargs,
    119     )
    120 else:
    121     new_value = bnb.nn.Params4bit(new_value, requires_grad=False, **kwargs).to(device)

File ~/anaconda3/envs/musai/lib/python3.11/site-packages/bitsandbytes/nn/modules.py:174, in Params4bit.from_prequantiz
ed(cls, data, quantized_stats, requires_grad, device, **kwargs)
    172 self = torch.Tensor._make_subclass(cls, data.to(device))
    173 self.requires_grad = requires_grad
--> 174 self.quant_state = QuantState.from_dict(qs_dict=quantized_stats, device=device)
    175 self.blocksize = self.quant_state.blocksize
    176 self.compress_statistics = self.quant_state.nested

File ~/anaconda3/envs/musai/lib/python3.11/site-packages/bitsandbytes/functional.py:626, in QuantState.from_dict(cls, 
qs_dict, device)
    623     qs_dict.update(unpack_tensor_to_dict(qs_dict.pop(qs_key)))
    625 qs_dict = {k.split('.')[-1]: v for k, v in qs_dict.items()}  # strip prefixes
--> 626 assert set(qs_dict.keys()).issubset(cls.valid_qs_keys)
    628 if 'nested_absmax' in qs_dict:
    629     offset = torch.tensor(float(qs_dict['nested_offset'])).to(device)

AssertionError: 
```

Any clue?

poedator commented:
Hi, @webpolis! Thank you for interest to this serialization method.
Apparently, there are some keys in `qs_dict.keys()` which are not expected by the model. 
I have 2 guesses:
- (most likely) your model used some draft naming scheme for the quantization components, that was changed later as this PR progressed. In this case, try to re-save the quntized model using the latest transformers (from github until 4.37 is out) and bnb>=0.41.3.
- something changed on bnb side (try with 0.41.3 or whatever worked)

webpolis commented:
Thanks @poedator .. that sounds reasonable. I'll try that.

hkhan9 commented:
Hi @poedator @amyeroberts thanks for releasing v4.37. I just installed it and saved the model using 4-bit. This was a huge achievement. Congratulations to the team !

Naozumi520 commented:
> Hi @poedator @amyeroberts thanks for releasing v4.37. I just installed it and saved the model using 4-bit. This was a huge achievement. Congratulations to the team !

May I have your script for 4-bit model export? Just used bnb_config and did not work. Model still as large as original.

Naozumi520 commented:
```py
from transformers import LlamaForCausalLM, AutoTokenizer, BitsAndBytesConfig
from peft import PeftModel
import torch

import argparse


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--base_model_name_or_path", type=str)
    parser.add_argument("--output_dir", type=str)
    parser.add_argument("--device", type=str, default="cpu")

    return parser.parse_args()


def main():
    args = get_args()

    if args.device == "auto":
        device_arg = {"device_map": "auto"}
    else:
        device_arg = {"device_map": {"": args.device}}

    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_use_double_quant=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_compute_dtype=torch.bfloat16,
    )

    print(f"Loading base model: {args.base_model_name_or_path}")

    model_4bit = LlamaForCausalLM.from_pretrained(
        args.base_model_name_or_path,
        low_cpu_mem_usage=True,
        load_in_4bit=True,
        return_dict=True,
        # quantization_config=bnb_config,
        torch_dtype=torch.float16,
        **device_arg,
    )

    print(f"Running 4bit")

    tokenizer = AutoTokenizer.from_pretrained(args.base_model_name_or_path)

    model_4bit.save_pretrained(f"{args.output_dir}")
    tokenizer.save_pretrained(f"{args.output_dir}")
    print(f"Model saved to {args.output_dir}")


if __name__ == "__main__":
    main()
    ```
